File: frontend/src/features/ortho/orthoSuperimposition.ts
import { api } from '../../services/api';

export interface SuperimpositionPixelROI {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface OrthoSuperimpositionSource {
  timepoint_id: number;
  timepoint_ordinal: number;
  occurred_at: string;
  cephalo_analysis_id: number;
  is_calibrated: boolean;
  mm_per_pixel: number | null;
}

export interface OrthoSuperimpositionContext {
  patient_id: number;
  ortho_case_id: number;
  from_source: OrthoSuperimpositionSource;
  to_source: OrthoSuperimpositionSource;
  quantitative_mm_allowed: boolean;
  applicability_status: string;
  method_id: string;
  method_version: string;
  quality_status: 'ENGINE_ESTIMATE_ONLY';
  clinically_validated: false;
}

export interface OrthoSuperimpositionRegistration {
  method_id: string;
  method_version: string;
  quality_status: 'ENGINE_ESTIMATE_ONLY';
  clinically_validated: false;
  transform_direction: 'moving_to_reference';
  matrix: [[number, number, number], [number, number, number]];
  rotation_degrees: number;
  uniform_scale: number;
  translation_px: { x: number; y: number };
  good_match_count: number;
  inlier_count: number;
  reference_roi: SuperimpositionPixelROI;
  moving_roi: SuperimpositionPixelROI;
  reference_size_px: { width: number; height: number };
  moving_size_px: { width: number; height: number };
  algorithm: Record<string, unknown>;
}

export interface OrthoSuperimpositionEstimate {
  context: OrthoSuperimpositionContext;
  registration: OrthoSuperimpositionRegistration;
}

export async function fetchOrthoSuperimpositionContext(
  patientId: number,
  caseId: number,
  fromTimepointId: number,
  toTimepointId: number,
): Promise<OrthoSuperimpositionContext> {
  const { data } = await api.get<OrthoSuperimpositionContext>(
    `/patients/${patientId}/ortho-case/${caseId}/superimposition/context`,
    { params: { from_timepoint_id: fromTimepointId, to_timepoint_id: toTimepointId } },
  );
  return data;
}

export async function estimateOrthoSuperimposition(
  patientId: number,
  caseId: number,
  payload: {
    from_timepoint_id: number;
    to_timepoint_id: number;
    reference_roi: SuperimpositionPixelROI;
    moving_roi: SuperimpositionPixelROI;
  },
): Promise<OrthoSuperimpositionEstimate> {
  const { data } = await api.post<OrthoSuperimpositionEstimate>(
    `/patients/${patientId}/ortho-case/${caseId}/superimposition/estimate`,
    payload,
  );
  return data;
}

File: backend/schemas/ortho_superimposition.py
from __future__ import annotations

import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class SuperimpositionPixelROI(BaseModel):
    model_config = ConfigDict(extra="forbid")
    x: int = Field(ge=0)
    y: int = Field(ge=0)
    width: int = Field(gt=0)
    height: int = Field(gt=0)


class OrthoSuperimpositionSourceOut(BaseModel):
    timepoint_id: int
    timepoint_ordinal: int
    occurred_at: datetime.datetime
    cephalo_analysis_id: int
    is_calibrated: bool
    mm_per_pixel: float | None = None


class OrthoSuperimpositionContextOut(BaseModel):
    patient_id: int
    ortho_case_id: int
    from_source: OrthoSuperimpositionSourceOut
    to_source: OrthoSuperimpositionSourceOut
    quantitative_mm_allowed: bool
    applicability_status: str
    method_id: str
    method_version: str
    quality_status: str
    clinically_validated: bool = False


class OrthoSuperimpositionEstimateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    from_timepoint_id: int
    to_timepoint_id: int
    reference_roi: SuperimpositionPixelROI
    moving_roi: SuperimpositionPixelROI


class OrthoSuperimpositionEstimateOut(BaseModel):
    context: OrthoSuperimpositionContextOut
    registration: dict[str, Any]
